uvdat/core/models/networks.py

```diff
@@ -0,0 +1,26 @@
+from django.contrib.gis.db import models as geo_models
+from django.db import models
+
+from .dataset import Dataset
+
+
+class NetworkNode(models.Model):
+    name = models.CharField(max_length=255, unique=True)
+    dataset = models.ForeignKey(Dataset, on_delete=models.CASCADE, related_name='network_nodes')
+    metadata = models.JSONField(blank=True, null=True)
+    capacity = models.IntegerField(null=True)
+    location = geo_models.PointField()
+    connected_edges = models.ManyToManyField('NetworkEdge')
+
+    def get_adjacent_nodes(self):
+        # TODO: get adjacent nodes from connected_edges
+        return []
+
+
+class NetworkEdge(models.Model):
+    name = models.CharField(max_length=255, unique=True)
+    dataset = models.ForeignKey(Dataset, on_delete=models.CASCADE, related_name='network_edges')
+    metadata = models.JSONField(blank=True, null=True)
+    capacity = models.IntegerField(null=True)
+    line = geo_models.LineStringField()
+    connected_nodes = models.ManyToManyField('NetworkNode')
```

Comment:
I think a many-to-many field is overkill for this, and could prevent us from enforcing correctness. It might just be simpler to do the following:

```suggestion
    from_node = models.ForeignKey(NetworkNode, related_name='+', on_delete=models.CASCADE)
    to_node = models.ForeignKey(NetworkNode, related_name='+', on_delete=models.CASCADE)
```

We could also store a field that tells us if the edge is directed or not. I don't know if that makes sense with any use cases we have thus far, but it's certainly a common use case w.r.t. network data:
```
directed = models.BooleanField(default=False)
```

Do you think this will work?

Reply:
Yes, I like this concept. I think that we should allow the backwards relation, though (not use "+" for the `related_name`). It would allow us to remove the `connected_edges` field on the `NetworkNode` model. That way it will always be edges referring to nodes.

Reply:
There's no reason a backwards relation is required. You can achieve the same outcome by just using the `NetworkEdge` model:
```python
NetworkEdge.objects.filter(from_node_id=1)
NetworkEdge.objects.filter(to_node_id=4)
```

The reason I opted to not use the backwards relation is because, from past experience, when you have two relations like that, it can get confusing what means what. For example, with the backwards relations like so:

```python
from_node = models.ForeignKey(NetworkNode, related_name='from_edges', on_delete=models.CASCADE)
to_node = models.ForeignKey(NetworkNode, related_name='to_edges', on_delete=models.CASCADE)
```

Let's say the edge table looks something like this:


| from_node  | to_node |
| ------------- | ------------- |
| 1 | 3 |
| 1 | 4 |
| 2  | 4 |

If you wanted to retrieve all of the edges in which node `1` is a _start node_, or wanted to know for which edges node `4` is an _end node_, you would do the following:

```python
NetworkNode.objects.get(id=1).from_edges.all()
NetworkNode.objects.get(id=4).to_edges.all()
```

I don't think it's readily apparent that `from_edges` returns "the edges for which the node is in the starting position". It would be very easy to think `from_edges` actually means "the edges _from_ other nodes into this node", i.e. "the edges in which this node is in the _ending_ position". Reversing the naming scheme has similar problems. (I actually had to re-write this, as the first time I mixed up the two).

I think the usage of `NetworkEdge` to achieve this is much more clear, and prevents confusion (it's also more simple, as now there's only one way to do this kind of query, instead of two or three).

What do you think? If you would still prefer a backwards relation, I think we should try to carefully name them, although I don't see a reason they're actually needed.

Reply:
Ok I see what you mean. I changed the `related_name`s to "+" in [3e6b67f](https://github.com/OpenGeoscience/uvdat/pull/30/commits/3e6b67f84f49ae6c126218a9d3b3d907a9a81c98).